src/App.tsx

```diff
@@ -1,223 +1,126 @@
-/* eslint-disable jsx-a11y/control-has-associated-label */
-import React, { useContext, useEffect, useRef } from 'react';
+import React, {
+  useContext, useEffect, useState, useMemo,
+} from 'react';
 import { AuthContext } from './components/Auth/AuthContext';
+import { TodoList } from './components/TodoList/TodoList';
+import { Filter } from './components/Filter/Filter';
+import { NewTodoField } from './components/NewTodoField/NewTodoField';
+import { Todo } from './types/Todo';
+import { Filters } from './types/Filters';
+import { getTodos } from './api/todos';
+import { ErrorNotification } from
+  './components/ErrorNotification/ErrorNotification';
+import { useAddingTodo } from './controllers/useAddingTodo';
+import { useDeletingTodos } from './controllers/useDeletingTodos';
+import { useError } from './controllers/useError';
+import { useUpdatingTodo } from './controllers/useUpdatingTodo';
 
 export const App: React.FC = () => {
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
   const user = useContext(AuthContext);
-  const newTodoField = useRef<HTMLInputElement>(null);
+
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [selectedStatus, setSelectedStatus] = useState(Filters.All);
+  const [loadingTodosIds, setLoadingTodosIds] = useState<number[]>([]);
+
+  const [errorMessage, showError, hideError] = useError();
 
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
+    if (user) {
+      getTodos(user.id)
+        .then(setTodos)
+        .catch(() => showError('Cannot load todos'));
     }
-  }, []);
+  }, [user]);
+
+  const activeTodosNumber = useMemo(() => (
+    todos.filter(todo => !todo.completed).length
+  ), [todos]);
+
+  const visibleTodos = useMemo(() => {
+    return todos.filter((todo) => {
+      switch (selectedStatus) {
+        case Filters.Active:
+          return !todo.completed;
+
+        case Filters.Completed:
+          return todo.completed;
+
+        case Filters.All:
+        default:
+          return todo;
+      }
+    });
+  }, [todos, selectedStatus]);
+
+  const completedTodosIds = useMemo(() => (
+    todos
+      .filter(todo => todo.completed)
+      .map(todo => todo.id)
+  ), [todos]);
+
+  const isAllCompletedTodos = todos.length === completedTodosIds.length;
+
+  const [isAddingTodo, temporaryNewTodo, addTodo] = useAddingTodo(
+    {
+      setTodos, showError,
+    },
+  );
+
+  const [deleteTodo, removeCompleted] = useDeletingTodos(
+    {
+      setLoadingTodosIds, setTodos, showError, completedTodosIds,
+    },
+  );
+
+  const [editTodo, handleToggleAll] = useUpdatingTodo(
+    {
+      setLoadingTodosIds, setTodos, showError, isAllCompletedTodos, todos,
+    },
+  );
```

Comment:
Wow 🔥 
Interesting approach, I wonder if mentors will have to say something about it.

Reply:
Thanks! I tried to play with custom hooks to see if it really improves code readability.